Replace `rz_core_cmd0` calls in `tui/panels.c` and port `cmd_help` commands to `RzShell` as `cmd_math`

brightprogrammer opened the pull request:
# SQUASH ME

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

- Replaces some of the `rz_core_cmd0` in `/core/tui/panels.c` with corresponding API/IPI calls as per the project (**Switch from command calls to the API**) requirements.
- Port `cmd_help` to RzShell.
- In lateast commits  `?` symbol used in `cmd_help` is changed to to `cmd_math` (new created command group) and `%` symbol. Most of the math commands now use `%` symbols instead of `?`.
- Some of the commands that made more sense in `cmd_shell` have been moved there, like the version info command or command excecution time calculator command.  

Closes https://github.com/rizinorg/rizin/issues/2834
Partially addresses https://github.com/rizinorg/rizin/issues/383

brightprogrammer committed:
Replaced some rz_core_cmd0 with corresponding API/IPI

ret2libc commented on librz/core/tui/panels.c:
```diff
@@ -3607,8 +3608,14 @@ int __writeValueCb(void *user) {
+		if (core->num->nc.errors) {
+			RZ_LOG_ERROR("Could not convert argument to number");
+			free(res);
+			return RZ_CMD_STATUS_ERROR;
```
you want to return something else here. Like `-1`

ret2libc commented on librz/core/tui/panels.c:
```diff
@@ -3607,8 +3608,14 @@ int __writeValueCb(void *user) {
+			return RZ_CMD_STATUS_ERROR;
+		}
 		free(res);
+		return rz_core_write_value_at(core, core->offset, value, 0) ? RZ_CMD_STATUS_OK : RZ_CMD_STATUS_ERROR;
```
```suggestion
		return rz_core_write_value_at(core, core->offset, value, 0) ? 0 : -1;
```

ret2libc commented on librz/core/cmd_descs/cmd_descs.h:
```diff
@@ -1374,6 +1374,8 @@ RZ_IPI RzCmdStatus rz_egg_reset_handler(RzCore *core, int argc, const char **arg
 RZ_IPI RzCmdStatus rz_egg_show_config_handler(RzCore *core, int argc, const char **argv);
 // "help"
 RZ_IPI RzCmdStatus rz_help_handler(RzCore *core, int argc, const char **argv);
+// "? <expr>"
+RZ_IPI void rz_help_calc_expr(RzCore* core, const char* input);
```
You don't want to place functions here. This is an auto-generated file as it should be written at the head of the file. Place this declaration in another header like rz_core.h

ret2libc commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,102 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 	}
 }
 
+RZ_IPI void rz_help_calc_expr(void* data, const char* input) {
```
```suggestion
RZ_API void rz_help_calc_expr(RzCore *core, const char* input) {
```

ret2libc replied:
Also, you may want to transform this into something like `rz_core_asm_plugins_print`, which accepts a `RzCmdStateOutput`.

Then you can call it with something like https://github.com/rizinorg/rizin/blob/dev/librz/core/cconfig.c#L426

brightprogrammer replied:
I used this function directly inside the `rz_cmd_help` function in `"? "` case. Would it be a good idea to modify it's signature? I'll be like copying the code? 

Since I needed only the part which calculates expression and prints it to screen, I removed that case from `rz_cmd_help` and created this new function and then used this function there directly.

XVilka commented on librz/core/tui/panels.c:
```diff
@@ -4238,7 +4245,7 @@ void __init_menu_saved_layout(void *_core, const char *parent) {
 	RzCore *core = (RzCore *)_core;
 	const char *color = core->cons->context->pal.graph_box2;
-	char *now = rz_core_cmd_str(core, "eco.");
+	char *now = rz_core_theme_get(core);
```
`const char*`

brightprogrammer replied:
can't do `const char*` because there's a split below.
https://github.com/rizinorg/rizin/blob/fb71f368596879eb1664c5c2ac04c0613f425a0f/librz/core/tui/panels.c#L4242

brightprogrammer replied:
I'll make a copy of string and pass it to this split.

brightprogrammer replied:
I changed this, just forgot to free the duplicated string in this commit, I'll push the change in next commit, marking this as resolved.

brightprogrammer committed:
Apply recommended fixes and replace some more rz_core_cmd calls

brightprogrammer committed:
Replace one more rz_core_cmd0 call

brightprogrammer commented on librz/core/canalysis.c:
```diff
@@ -6393,3 +6393,34 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
 	return start_addr;
 }
+
+RZ_API void rz_core_perform_auto_analysis(RzCore *core, RzCoreAnalysisType type) {
```
had to move this function and related enum outside to this file in order to be able to use it directly instead of calling `rz_core_cmd("aa")` or `rz_core_cmd("aaa")` or `rz_core_cmd("aaaa")`.

wargio replied:
that sounds like a good idea.

brightprogrammer commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -5894,55 +5898,18 @@ RZ_IPI RzCmdStatus rz_analysis_basic_block_find_paths_handler(RzCore *core, int
 	return status;
 }
 
-typedef enum {
```
enum moved to `rz_core.h`

wargio commented on librz/core/canalysis.c:
```diff
@@ -6393,3 +6393,34 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
 	return start_addr;
 }
+
+RZ_API void rz_core_perform_auto_analysis(RzCore *core, RzCoreAnalysisType type) {
```
before it was not an `RZ_API` so there was no need for this but you need to add these changes:
```suggestion
RZ_API void rz_core_perform_auto_analysis(RZ_NONNULL RzCore *core, RzCoreAnalysisType type) {
	rz_return_if_fail(core);
```

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,102 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 	}
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RzCore* data, const char* input) {
```
same thing here (as above)

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,102 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RzCore* data, const char* input) {
+	RzCore* core = (RzCore*)data;
```
why you need to cast something that is `RzCore` already into `RzCore` ? just rename `data` argument into `core` and delete this line.

brightprogrammer replied:
I'm sorry, this was a side effect of copy pasting and forgot to change that. Will be changed in upcoming commit.

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,102 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+	double d;
+	float f;
+	char number[128], out[128] = RZ_EMPTY;
+	char *const inputs = strdup(input + 1);
```
```suggestion
	const char *inputs = strdup(input + 1);
```

brightprogrammer replied:
this results in meson build failures for some compilers... Built on my system with a warning. Maybe I should remove const qualifier, because a function expects the input in that format
![image](https://user-images.githubusercontent.com/37365905/224626108-3716e0df-99a1-4b04-b24a-4659665c1350.png)

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,102 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+		}
+		ut64 n = rz_num_math(core->num, str);
+		if (core->num->dbz) {
+			RZ_LOG_ERROR("core: RzNum ERROR: Division by Zero\n");
```
```suggestion
			RZ_LOG_ERROR("core: RzNum ERROR: Division by Zero\n");
			return;
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -717,6 +723,8 @@ RZ_API bool rz_core_analysis_hint_set_offset(RZ_NONNULL RzCore *core, RZ_NONNULL
 RZ_API bool rz_core_analysis_continue_until_syscall(RZ_NONNULL RzCore *core);
 RZ_API bool rz_core_analysis_continue_until_call(RZ_NONNULL RzCore *core);
 
+RZ_API void rz_core_perform_auto_analysis(RzCore *core, RzCoreAnalysisType type);
```
```suggestion
RZ_API void rz_core_perform_auto_analysis(RZ_NONNULL RzCore *core, RzCoreAnalysisType type);
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -1266,6 +1274,12 @@ typedef bool (*RzCmdRegSync)(RzCore *core, RzRegisterType type, bool write);
 
+/* rz_core_cmd replacement function declarations */
+// "? <expr>"
+RZ_API void rz_core_cmd_help_calc_expr(RzCore* core, const char* input);
```
```suggestion
RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore* core, RZ_NONNULL const char* input);
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -1266,6 +1274,12 @@ typedef bool (*RzCmdRegSync)(RzCore *core, RzRegisterType type, bool write);
+// "? <expr>"
+RZ_API void rz_core_cmd_help_calc_expr(RzCore* core, const char* input);
+// a? or aaaaa like help message cases
+RZ_API void rz_core_cmd_show_analysis_help(RzCore* core);
```
```suggestion
RZ_API void rz_core_cmd_show_analysis_help(RZ_NONNULL RzCore* core);
```

wargio reviewed (changes requested):
For each of `RZ_NONNULL` added to the arguments, please add (if is not there) a `rz_return_if_fail(cond);` if the function returns `void` or `rz_return_val_if_fail(cond, value);` if returns anything else but `void`.

brightprogrammer committed:
Make requested changes/fixes

wargio commented on librz/core/canalysis.c:
```diff
@@ -6394,7 +6394,9 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
 
-RZ_API void rz_core_perform_auto_analysis(RzCore *core, RzCoreAnalysisType type) {
+RZ_API void rz_core_perform_auto_analysis(RZ_NONNULL RzCore *core, RzCoreAnalysisType type) {
+	rz_return_if_fail(core != NULL);
```
```suggestion
	rz_return_if_fail(core);
```

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,14 +430,15 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
-RZ_API void rz_core_cmd_help_calc_expr(RzCore* data, const char* input) {
-	RzCore* core = (RzCore*)data;
+RZ_API void rz_core_cmd_help_calc_expr(RzCore* core, const char* input) {
+	rz_return_if_fail(core != NULL);
```
```suggestion
	rz_return_if_fail(core);
```

wargio commented on librz/include/rz_core.h:
```diff
@@ -1276,9 +1276,9 @@ RZ_API RZ_OWN RzList /*<RzRegItem *>*/ *rz_core_reg_filter_items_sync(RZ_NONNULL
 /* rz_core_cmd replacement function declarations */
 // "? <expr>"
-RZ_API void rz_core_cmd_help_calc_expr(RzCore* core, const char* input);
+RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore* core, const char* input);
```
input must be also `RZ_NONNULL` and also on the C file

brightprogrammer committed:
Make suggested fixes and run clang-format

brightprogrammer committed:
Minor change

brightprogrammer commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -2028,7 +2028,7 @@ RZ_IPI RzCmdStatus rz_analysis_global_variable_retype_handler(RzCore *core, int
 
 RZ_API void rz_core_cmd_show_analysis_help(RzCore *core) {
-	rz_return_if_fail(core != NULL);
+	rz_return_if_fail(core);
```
forgot to change this and this commit is rebased with dev (as guided in `CONTRIBUTING.md`)

XVilka replied:
Could you make it `RZ_IPI`?

brightprogrammer replied:
this is used in `librz/main` module
![image](https://user-images.githubusercontent.com/37365905/226920383-8abc7993-564b-4ef8-b20c-a184b7768439.png)

ret2libc commented on librz/main/rizin.c:
```diff
@@ -851,12 +851,14 @@ RZ_API int rz_main_rizin(int argc, const char **argv) {
 		}
-		rz_core_cmd0(r, "R!=");
+		// rz_core_cmd0(r, "R!=");
+		RZ_FREE(r->cmdremote);
+		r->cmdremote = rz_str_trim_dup("0");
```
Please extract this into `rz_core_rtr_enable(RZ_NONNULL RzCore *, RZ_NONNULL char *cmdremote)` and use it in both cmd_remote.c and here.

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1266,6 +1274,12 @@ typedef bool (*RzCmdRegSync)(RzCore *core, RzRegisterType type, bool write);
 RZ_API bool rz_core_reg_assign_sync(RZ_NONNULL RzCore *core, RZ_NONNULL RzReg *reg, RzCmdRegSync sync_cb, RZ_NONNULL const char *name, ut64 val);
 RZ_API RZ_OWN RzList /*<RzRegItem *>*/ *rz_core_reg_filter_items_sync(RZ_NONNULL RzCore *core, RZ_NONNULL RzReg *reg, RzCmdRegSync sync_cb, RZ_NULLABLE const char *filter);
 
+/* rz_core_cmd replacement function declarations */
```
no need for this, I think

ret2libc commented on librz/include/rz_core.h:
```diff
@@ -1266,6 +1274,12 @@ typedef bool (*RzCmdRegSync)(RzCore *core, RzRegisterType type, bool write);
 RZ_API RZ_OWN RzList /*<RzRegItem *>*/ *rz_core_reg_filter_items_sync(RZ_NONNULL RzCore *core, RZ_NONNULL RzReg *reg, RzCmdRegSync sync_cb, RZ_NULLABLE const char *filter);
 
+/* rz_core_cmd replacement function declarations */
+// "? <expr>"
```
remove these comments as well.

ret2libc reviewed (commented):
Nice! Just few minor comments, but seems good to me :) Thanks!

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,105 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RzCore *core, const char *input) {
+	rz_return_if_fail(core);
+	rz_return_if_fail(input);
```
```suggestion
	rz_return_if_fail(core && input);
```

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,105 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 	}
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RzCore *core, const char *input) {
```
```suggestion
RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore *core, RZ_NONNULL const char *input) {
```

wargio commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -2027,6 +2027,11 @@ RZ_IPI RzCmdStatus rz_analysis_global_variable_retype_handler(RzCore *core, int
 	return RZ_CMD_STATUS_OK;
 }
 
+RZ_API void rz_core_cmd_show_analysis_help(RzCore *core) {
```
```suggestion
RZ_API void rz_core_cmd_show_analysis_help(RZ_NONNULL RzCore *core) {
```

wargio commented on librz/core/canalysis.c:
```diff
@@ -6384,3 +6384,36 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
 	return start_addr;
 }
+
+RZ_API void rz_core_perform_auto_analysis(RzCore *core, RzCoreAnalysisType type) {
```
```suggestion
RZ_API void rz_core_perform_auto_analysis(RZ_NONNULL RzCore *core, RzCoreAnalysisType type) {
```

brightprogrammer committed:
Make recommended changes

brightprogrammer committed:
Make recommended changes

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,104 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 	}
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RzCore *core, const char *input) {
```
Since you are modifying this code, I suggest porting this particular command to the rzshell as well, and use `RzCmdStateOutput`, just add corresponding code into `librz/core/cmd_descs/cmd_help.yaml`. And extract evaluation function to the API, while keeping printing function as the IPI.

brightprogrammer replied:
ok :smile:

brightprogrammer committed:
Make recommended changes

wargio commented on librz/include/rz_core.h:
```diff
@@ -1274,11 +1274,9 @@ typedef bool (*RzCmdRegSync)(RzCore *core, RzRegisterType type, bool write);
 RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore *core, RZ_NONNULL const char *input);
-// a? or aaaaa like help message cases
 RZ_API void rz_core_cmd_show_analysis_help(RZ_NONNULL RzCore *core);
+RZ_API void rz_core_rtr_enable(RZ_NONNULL RzCore *core, RZ_BORROW const char *cmdremote);
```
RZ_BORROW is only if the buffer is allocated by another function and held by a structure which should not be handled by the user (i.e. freed).

XVilka replied:
Yes, please remove it

brightprogrammer committed:
Port help commands to rzshell

brightprogrammer commented:
I haven't written command handlers yet. That will be done in next two to three days :smile: Please review `cmd_help.yaml` till then. I'll be going through each entry again when implementing the handlers because I was confused about how arguments work in rzshell handlers. Will figure that out while working...

wargio commented on librz/core/cmd_descs/cmd_help.yaml:
```diff
@@ -0,0 +1,396 @@
+# SPDX-FileCopyrightText: 2021 RizinOrg <info@rizin.re>
```
```suggestion
# SPDX-FileCopyrightText: 2023 RizinOrg <info@rizin.re>
```

wargio commented on librz/core/cmd_descs/cmd_help.yaml:
```diff
@@ -0,0 +1,396 @@
+      - name: "???"
+        summary: "Display ? Help"
+        cname: show_help_question
+      - name: "????"
+        summary: "What are you doing?"
+        cname: what_are_you_doing?
```
remove that function and delete this command.

wargio reviewed (changes requested):
Just a couple of changes and then can be merged (once you implement the handlers).

XVilka commented on librz/core/canalysis.c:
```diff
@@ -6384,3 +6384,36 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
 	return start_addr;
 }
+
+RZ_API void rz_core_perform_auto_analysis(RZ_NONNULL RzCore *core, RzCoreAnalysisType type) {
```
Please add Doxygen documentation

XVilka commented on librz/core/canalysis.c:
```diff
@@ -6384,3 +6384,36 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
+	}
+	// Run pending analysis immediately after analysis
+	// Usefull when running commands with ";" or via rizin -c,-i
+	debugger = core->dbg->cur ? strdup(core->dbg->cur->name) : strdup("esil");
```
You can move definition of the variable here, no need to define everything at the beginning of the function - we support only >=C99 standard.

brightprogrammer replied:
Moving declaration here shows some compiler warning for non initialization of `debugger`. I'll need some time on that because I was focusing on rzshell.

wargio replied:
but `char *debugger = NULL;` at line 6398, so it should not.

XVilka replied:
Don't forget this please

XVilka commented on librz/core/cmd/cmd_analysis.c:
```diff
@@ -5894,55 +5899,18 @@ RZ_IPI RzCmdStatus rz_analysis_basic_block_find_paths_handler(RzCore *core, int
-
 RZ_IPI RzCmdStatus rz_analyze_simple_handler(RzCore *core, int argc, const char **argv) {
-	core_perform_auto_analysis(core, CORE_ANALYSIS_SIMPLE);
+	rz_core_perform_auto_analysis(core, RZ_CORE_ANALYSIS_SIMPLE);
```
For this new function - `rz_core_perform_auto_analysis()` I recommend to create a new unit/integration test, that will try all levels of the analysis for some target and compare the results between them - e.g. that some functions are missing on SIMPLE level but present on more advanced. You can see these as an example:
- https://github.com/rizinorg/rizin/blob/dev/test/integration/test_open_analyse_save_load_project.c
- https://github.com/rizinorg/rizin/blob/dev/test/integration/test_analysis_global_var.c

No need to save/load project, just opening file multiple times for every different analysis will suffice.

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,6 +430,104 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
 	}
 }
 
+RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore *core, RZ_NONNULL const char *input) {
```
1. No need for it to be and API, IPI will suffice.
2. I recommend to use rzshell `intc argc, char **argv`-style arguments, also with `RzCmdStateOutput` for using JSON and other kinds of output

XVilka replied:
Just make it `IPI` for this PR, it can be converted later then.

wargio replied:
set this as `RZ_IPI` and move the header definition under `core_private.h`

XVilka commented on librz/core/cmd/cmd_remote.c:
```diff
@@ -253,3 +255,10 @@ RZ_IPI RzCmdStatus rz_remote_tcp_handler(RzCore *core, int argc, const char **ar
 	return RZ_CMD_STATUS_ERROR;
 }
+
+RZ_API void rz_core_rtr_enable(RZ_NONNULL RzCore *core, RZ_BORROW const char *cmdremote) {
```
Maybe no need for it to be an API?

brightprogrammer committed:
cmd_help port to rzshell complete

brightprogrammer commented on librz/core/cmd/cmd_help.c:
```diff
@@ -529,6 +435,7 @@ RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore *core, RZ_NONNULL const
 }
 
 RZ_IPI int rz_cmd_help(void *data, const char *input) {
```
Left out commented part are commands that were not ported. For `@`, `@@` and `@@@` the generation script is generating wrong output so compiler shows compile error.

This function is kept not deleted itself because it's used in `cmd.c` @ line 5597. I tried commenting that out but compiler again shows some errors. I can figure that out later. For now it's just this. There are also some other minor changes left out. Waiting your reviews! :smile:

brightprogrammer commented on librz/core/cmd/cmd_help.c:
```diff
@@ -805,297 +462,6 @@ RZ_IPI int rz_cmd_help(void *data, const char *input) {
-		break;
-	}
 	case '?': // "??"
 		if (input[1] == '?') {
```
Some of the commands like `??` work differently now. Previously `??x` would have executed without any errors but now it's `?? x` (notice the SPACE in between)

brightprogrammer committed:
run clang-format

brightprogrammer committed:
small segfault fix

brightprogrammer committed:
more fixes

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -690,7 +690,7 @@ RZ_IPI RzCmdStatus rz_print_boundaries_prot_handler(RzCore *core, int argc, cons
 
 RZ_IPI RzCmdStatus rz_print_djb2_hash_handler(RzCore *core, int argc, const char **argv) {
-	for (int i = i; i < argc; i++) {
+	for (int i = 1; i < argc; i++) {
```
nice catch. this means there is no test for this.

brightprogrammer replied:
should I write test for this?

XVilka replied:
Would be nice, yes.

brightprogrammer committed:
fix code and modify cmd tests

brightprogrammer committed:
added test of new analysis function

brightprogrammer committed:
added new test case to meson build system

brightprogrammer committed:
meson.build fix

brightprogrammer committed:
fix

brightprogrammer committed:
test fix

brightprogrammer committed:
test fix

brightprogrammer committed:
mistake fix

brightprogrammer committed:
fix left out tests

brightprogrammer committed:
Fix tests

brightprogrammer committed:
Fix failed test

brightprogrammer committed:
Fix failed test

brightprogrammer committed:
Fix indentation problem

wargio commented on test/db/tools/rz:
```diff
@@ -74,7 +74,7 @@ FILE==
 EXPECT=<<EOF
 1
-EOF
+EO
```
this is wrong.
```suggestion
EOF
```

brightprogrammer replied:
This was fixed in last commits

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -857,32 +857,24 @@ RZ_IPI RzCmdStatus rz_compare_and_set_core_num_value_handler(RzCore *core, int a
-		if (core->num->value == 0) {
-			core->num->value = rz_core_cmd(core, argv[1], 0);
-		}
+	if (core->num->value == 0) {
```
```suggestion
	if (!core->num->value) {
```

brightprogrammer replied:
This command will then do the same thing as `?+` That's why I changed it to check for zero value.

wargio replied:
doesn't matter. when checking for `if zero` or `if NULL` just do the comparison as `if (!variable)`

brightprogrammer replied:
Ok

wargio commented on test/db/cmd/cmd_help:
```diff
@@ -58,9 +58,9 @@ CMDS=<<EOF
-?-
+#?!
+#?+
+#?-
```
why commenting?

brightprogrammer replied:
These commands print help when wrong arguments (or no arguments) are passed instead of just printing the `core->num->value`. We have `?v` command group for that.

brightprogrammer replied:
These checks were there to test whether they print the `core->num->value` or not

brightprogrammer committed:
Add btw command and change a command handler

wargio commented on librz/core/cmd/cmd_help.c:
```diff
@@ -434,63 +435,8 @@ RZ_API void rz_core_cmd_help_calc_expr(RZ_NONNULL RzCore *core, RZ_NONNULL const
-			break;
-		}
-		*/
+	// TODO: I don't know how to remove this function.
+	// It's called in cmd.c file in rz_core_cmd_init
```
open an issue for this

brightprogrammer replied:
I tried removing it this time and it worked. Maybe it failed before because not all commands were ported.

brightprogrammer committed:
Remove rz_cmd_help completely

XVilka commented on librz/core/tui/panels.c:
```diff
@@ -5188,7 +5196,8 @@ bool __handle_console(RzCore *core, RzPanel *panel, const int key) {
 		if (*cmd) {
 			if (!strcmp(cmd, "clear")) {
-				rz_core_cmd0(core, ":>$console");
+				// rz_core_cmd0(core, ":>$console");
```
Remove

XVilka commented on librz/main/rizin.c:
```diff
@@ -1221,7 +1221,13 @@ RZ_API int rz_main_rizin(int argc, const char **argv) {
+		// initalize io subsystem
+		char *res = rz_io_system(r->io, NULL);
+		if (res) {
+			rz_cons_printf("%s\n", res);
```
Could just use `rz_cons_println(res)`

XVilka commented on test/db/cmd/feat_input:
```diff
@@ -78,7 +78,7 @@ RUN
 NAME=at3 "?vi $$"@3;"?e bar"
 FILE=malloc://1024
-CMDS="?vi $$"@ 3;"?e bar"
+CMDS=?vi $$ @ 3; ?e bar
```
Please change the name of the tests as well, to reflect the changes

XVilka commented on test/integration/test_auto_analysis.c:
```diff
@@ -0,0 +1,58 @@
+	ut32 experimental_fcn_list_length = rz_list_length(experimental_fcn_list);
+	eprintf("functions count = %d\n", experimental_fcn_list_length);
+
+	// again more in experimental
```
I recommend also checking for some particular functions presence, not just simple count comparison.
Moreover, I recommend adding also ARM binary as a second test here, then you could extract the shared code in the static function and run over multiple different files.

XVilka commented on test/integration/test_auto_analysis.c:
```diff
@@ -0,0 +1,58 @@
+	// not true for all binaries but for most it must be true
+	mu_assert("deep analysis success", deep_fcn_list_length >= simple_fcn_list_length);
+
+	// aaa
```
`aaaa`

XVilka commented on librz/include/rz_core.h:
```diff
@@ -680,6 +680,12 @@ typedef struct rz_core_analysis_name_t {
 } RzCoreAnalysisName;
 
+typedef enum {
+	RZ_CORE_ANALYSIS_SIMPLE, ///< aa
```
It needs better description, at least some context

wargio commented on librz/core/canalysis.c:
```diff
@@ -6384,3 +6384,41 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
+	}
+	// Run pending analysis immediately after analysis
+	// Usefull when running commands with ";" or via rizin -c,-i
+	debugger = core->dbg->cur ? strdup(core->dbg->cur->name) : strdup("esil");
+	if (core->io && core->io->desc && core->io->desc->plugin && !core->io->desc->plugin->isdbg) {
+		// set it only if is debugging
+		RZ_FREE(debugger);
+	}
+	rz_cons_clear_line(1);
+	rz_core_analysis_everything(core, type == RZ_CORE_ANALYSIS_EXPERIMENTAL, debugger);
```
```suggestion
	if (!(core->io && core->io->desc && core->io->desc->plugin && !core->io->desc->plugin->isdbg)) {
		// set it only if is debugging
		debugger = core->dbg->cur ? strdup(core->dbg->cur->name) : strdup("esil");
	}
	rz_cons_clear_line(1);
	rz_core_analysis_everything(core, type == RZ_CORE_ANALYSIS_EXPERIMENTAL, debugger);
```

wargio commented on librz/core/canalysis.c:
```diff
@@ -6384,3 +6384,41 @@ RZ_API ut64 rz_core_prevop_addr_force(RzCore *core, ut64 start_addr, int numinst
+	}
+	// Run pending analysis immediately after analysis
+	// Usefull when running commands with ";" or via rizin -c,-i
+	debugger = core->dbg->cur ? strdup(core->dbg->cur->name) : strdup("esil");
+	if (core->io && core->io->desc && core->io->desc->plugin && !core->io->desc->plugin->isdbg) {
+		// set it only if is debugging
+		RZ_FREE(debugger);
```
```suggestion
	if (core->io && core->io->desc && core->io->desc->plugin && core->io->desc->plugin->isdbg) {
		// set it only if is debugging
		debugger = core->dbg->cur ? strdup(core->dbg->cur->name) : strdup("esil");
```

brightprogrammer committed:
Removed unnecessary functions and improved auto_analysis test

brightprogrammer committed:
Run clang-format

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+
+RZ_IPI RzCmdStatus rz_print_binary_handler(RzCore *core, int argc, const char **argv) {
+	char out[128] = RZ_EMPTY;
+	ut64 n;
```
fold these two lines into one

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+}
+
+RZ_IPI RzCmdStatus rz_print_bitstring_handler(RzCore *core, int argc, const char **argv) {
+	ut64 n = 0;
```
Ssame, fold

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+
+RZ_IPI RzCmdStatus rz_show_value_handler(RzCore *core, int argc, const char **argv) {
+	ut64 n;
+	if (argc == 1) {
```
You can use `ut64 n = argc == 1 ? core->num->value : rz_num_math()`

XVilka replied:
Though I would use `argc < 2` to show the logic better instead of direct comparison.

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+
+RZ_IPI RzCmdStatus rz_show_value_hex_handler(RzCore *core, int argc, const char **argv) {
+	ut64 n;
+	if (argc == 1) {
```
Same

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+
+RZ_IPI RzCmdStatus rz_show_value_i1_handler(RzCore *core, int argc, const char **argv) {
+	ut64 n;
+	if (argc == 1) {
```
Same, etc

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+	return RZ_CMD_STATUS_OK;
+}
+
+/* RZ_IPI RzCmdStatus rz_show_help_at_handler(RzCore *core, int argc, const char **argv) { */
```
Remove

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+}
+
+RZ_IPI RzCmdStatus rz_echo_msg_newline_handler(RzCore *core, int argc, const char **argv) {
+	if (argc != 1) {
```
`argc > 1` probably better

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,764 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+}
+
+RZ_IPI RzCmdStatus rz_phys2virt_handler(RzCore *core, int argc, const char **argv) {
+	ut64 n = argc == 2 ? rz_num_math(core->num, argv[1]) : core->offset;
```
Just like you did here.

XVilka commented on test/db/cmd/cmd_help:
```diff
@@ -47,34 +47,25 @@ RUN
 ?v
-?=
-??
+# ?=
```
So what do we do with those? Remove them maybe? They aren't valid commands in the rzshell anyways, since require additional argument, I suppose?

XVilka replied:
Don't forget these

XVilka reviewed (changes requested):
Overall is good, see few minor changes necessary though.

brightprogrammer committed:
Minor changes in cmd_help.c

XVilka commented on librz/core/cmd/cmd_help.c:
```diff
@@ -430,696 +336,722 @@ RZ_IPI void rz_core_clippy_print(RzCore *core, const char *msg) {
+RZ_IPI RzCmdStatus rz_generate_random_number_handler(RzCore *core, int argc, const char **argv) {
+	// TODO : Add support for 64bit random numbers
+	ut64 b = 0;
+	ut32 r = UT32_MAX;
```
Could be folded with the following 3 lines as well using `?:` operator.

XVilka reviewed (changes requested):
Few last nitpicks, apart from that - LGTM. Please also rebase on top of the latest `dev`.

brightprogrammer committed:
Requested minor changes

XVilka reviewed (approved):
From the code PoV looks good. @ret2libc please check if all commands in YAML make sense.

ret2libc reviewed (changes requested):
There is a problem with this PR for now: How do you list all commands under `?`? So, I remember we discussed whether we wanted to keep the `?` prefix for these commands or actually use something else, considering that `?` is already used as the suffix to list sub-commands (e.g. `a?` list sub-commands of `a`).

@thestr4ng3r @XVilka @wargio @kazarmy @ITAYC0HEN WDYT?

brightprogrammer commented:
There was a command ??? To show help of all ? based commands but I had to remove it because someone requested it...

ret2libc commented on test/integration/test_auto_analysis.c:
```diff
@@ -0,0 +1,130 @@
+	{"bins/elf/ls", "fcn.00016320", RZ_CORE_ANALYSIS_DEEP},
+};
+
+const ut32 bin_fcn_list_size = sizeof(bin_fcn_list)/sizeof(bin_fcn_list[0]);
```
Use `RZ_ARRAY_SIZE`

ret2libc commented on test/integration/test_auto_analysis.c:
```diff
@@ -0,0 +1,130 @@
+}
+
+bool test_new_fcn_detected() {
+	for(ut32 i = 0; i < bin_fcn_list_size; ++i) {
```
this is quite slow as it re-analyze the same binary multiple times for each function, but ok for now 👍

brightprogrammer replied:
how can I improve this test?

XVilka replied:
You could just make this array two-dimensional. One dimension are files, another dimension are functions/analysis levels.

ret2libc commented on librz/core/cmd_descs/cmd_help.yaml:
```diff
@@ -0,0 +1,559 @@
+            arg_str: " 12 Hello"
+            comment: "LL (00110b selected : big-endian bitstring)"
+  - name: "?o"
+    summary: Evalue expression and print value in octal.
```
```suggestion
    summary: Evaluate expression and print value in octal.
```

ret2libc commented on librz/core/cmd_descs/cmd_help.yaml:
```diff
@@ -0,0 +1,559 @@
+        summary: Show patch version
+        cname: show_version_patch
+        args: []
+  - name: "?l"
+    summary: String length calculation commands
+    subcommands:
+      - name: "?l"
+        summary: Calculate length of given string and store in $? register
+        cname: compute_string_length
+        args:
+          - name: str
+            type: RZ_CMD_ARG_TYPE_STRING
+      - name: "?lq"
+        summary: Calculate length of given string and store in $? register but print nothing (quietly).
+        cname: compute_string_length_quiet
+        args:
+          - name: str
+            type: RZ_CMD_ARG_TYPE_STRING
```
for these kinds of commands you can use something like https://github.com/rizinorg/rizin/blob/dev/librz/core/cmd_descs/cmd_write.yaml#L318 . You just have to specify both the STANDARD and QUIET mode and then implement them in the same handler (like https://github.com/rizinorg/rizin/blob/dev/librz/core/cio.c#L802).

brightprogrammer committed:
Change from different hanlders for quite mode to RzCmdStateOutput method

XVilka commented:
> There is a problem with this PR for now: How do you list all commands under `?`? So, I remember we discussed whether we wanted to keep the `?` prefix for these commands or actually use something else, considering that `?` is already used as the suffix to list sub-commands (e.g. `a?` list sub-commands of `a`).
> 
> @thestr4ng3r @XVilka @wargio @kazarmy @ITAYC0HEN WDYT?

I propose reusing `%` for these math evaluation commands. Currently `%` is used to access and set environment variables which is not so common use case, unlike evaluating something, also there is already `env` command for that. Thus, removing `%` for environment variables and using it for math expressions.

Moreover, `%` character is somewhat loosely related to the "mathematics" and "evaluation" semantically.

For commands like `?e` we can remove them, since there is already `echo`. If some non-math and non-evaluation command is there - it's better to move to the `shell` category, e.g. `time` instead of `?t`

wargio commented:
i'm ok with `%`, but it is used currently for environment vars.
i propose instead `^` which is not used

brightprogrammer commented:
If we eventually need to move non mathematical commands out of this group, then I guess what ret2libc suggested is a good idea. We can use `m` for mathematical expression and `m` isn't used. Then other commands can be moved to a new misc command group by `^` or `%` or any other symbol not in use.

But I guess that should be done in another PR because that'd mean changing tests too.

brightprogrammer committed:
Improve auto analysis integration test

ret2libc commented:
> If we eventually need to move non mathematical commands out of this group, then I guess what ret2libc suggested is a good idea. We can use `m` for mathematical expression and `m` isn't used. Then other commands can be moved to a new misc command group by `^` or `%` or any other symbol not in use.
> 
> But I guess that should be done in another PR because that'd mean changing tests too.

The problem is that right now this can't be merged with `?` because `???` does not work.

brightprogrammer committed:
Change from question_mark to percent for math and add cmd_math

brightprogrammer committed:
Fix cmd_descs_generate.py and patch rizin-shell-parser grammmar

Fixes for failed tests and reviewed changes

	modified:   ../librz/util/hex.c
	modified:   ../test/db/analysis/ppc
	modified:   ../test/db/analysis/x86_64
	modified:   ../test/db/archos/linux-x64/dbg_bps
	modified:   ../test/db/archos/linux-x64/dbg_dr
	modified:   ../test/db/archos/not-windows-any/cmd_interpreter
	modified:   ../test/db/formats/dmp/dmp
	modified:   ../test/db/formats/elf/pie
	modified:   ../test/db/formats/pdb
	modified:   ../test/db/io/neg
	modified:   ../test/db/tools/rz_ax
	modified:   ../test/db/tools/rz_bin

XVilka commented:
```
pylint: Command line or configuration file:1: UserWarning: Specifying exception names in the overgeneral-exceptions option without module name is deprecated and support for it will be removed in pylint 3.0. Use fully qualified name (maybe 'builtins.BaseException' ?) instead.
pylint: Command line or configuration file:1: UserWarning: Specifying exception names in the overgeneral-exceptions option without module name is deprecated and support for it will be removed in pylint 3.0. Use fully qualified name (maybe 'builtins.Exception' ?) instead.
************* Module cmd_descs_generate
librz/core/cmd_descs/cmd_descs_generate.py:188:8: W0[7](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:8)19: Raising too general exception: Exception (broad-exception-raised)
librz/core/cmd_descs/cmd_descs_generate.py:731:[8](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:9): W071[9](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:10): Raising too general exception: Exception (broad-exception-raised)
************* Module fuzz_rz_asm
test/fuzz/scripts/fuzz_rz_asm.py:85:[11](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:12): R1735: Consider using '{"cause": cause, "ins": ins, "inpairs": inpairs, "oins": oins, "case": inskey, ... }' instead of a call to 'dict'. (use-dict-literal)

-----------------------------------
```

brightprogrammer commented:
> ```
> pylint: Command line or configuration file:1: UserWarning: Specifying exception names in the overgeneral-exceptions option without module name is deprecated and support for it will be removed in pylint 3.0. Use fully qualified name (maybe 'builtins.BaseException' ?) instead.
> pylint: Command line or configuration file:1: UserWarning: Specifying exception names in the overgeneral-exceptions option without module name is deprecated and support for it will be removed in pylint 3.0. Use fully qualified name (maybe 'builtins.Exception' ?) instead.
> ************* Module cmd_descs_generate
> librz/core/cmd_descs/cmd_descs_generate.py:188:8: W0[7](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:8)19: Raising too general exception: Exception (broad-exception-raised)
> librz/core/cmd_descs/cmd_descs_generate.py:731:[8](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:9): W071[9](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:10): Raising too general exception: Exception (broad-exception-raised)
> ************* Module fuzz_rz_asm
> test/fuzz/scripts/fuzz_rz_asm.py:85:[11](https://github.com/rizinorg/rizin/actions/runs/4546398351/jobs/8020421131?pr=3421#step:5:12): R1735: Consider using '{"cause": cause, "ins": ins, "inpairs": inpairs, "oins": oins, "case": inskey, ... }' instead of a call to 'dict'. (use-dict-literal)
> 
> -----------------------------------
> ```

How do I fix this?

XVilka commented:
@brightprogrammer:
1. Make a commit fixing these files (you can run `pylint` and `prettier` locally)
2. Push it here, if everything is green - send a separate PR with those fixes, wait for it to pass all tests, we will merge if everything is fine
3. Rebase this PR

ret2libc commented on test/db/tools/rz_ax:
```diff
@@ -354,7 +354,7 @@ RUN
 NAME=Python string '\x33\x44\x88'
 FILE==
-CMDS=echo "foo = '\x33\x44\x88'" | rz-ax -F
+CMDS=echo foo = "'\x33\x44\x88'" | rz-ax -F
```
is this required? I thought `echo "foo = '\x33\x44\x88'"` would work too.

brightprogrammer replied:
It makes more sense to be like this. There are two python tests for different python string syntax.
- `"string"` which is tested by `'"string"'`
- `'string'` which is tested by `"'string'"`

ret2libc commented on test/db/esil/x86_32:
```diff
@@ -188,9 +188,9 @@ aes
-ar?edi
-ar?esi
-ar?ecx
+arechodi
+arechosi
+arechocx
```
wrong replacement

ret2libc commented on test/db/cmd/feat_grep:
```diff
@@ -165,10 +165,10 @@ i~:..3
-?e 0\n1\n2\n3\n4~:-2..
-?e 0\n1\n2~:0..
-?e 0\n1\n2~:..
+echo "0\n1\n2\n3\n4"~:-2
+echo "0\n1\n2\n3\n4"~:-2..
+echo "0\n1\n2"~:0..
+echo "0\n1\n2"~:..
```
TODO: check if `echo` unescape characters like `?e` did.

brightprogrammer replied:
~~No that doesn't work, because I replaced the old echo code with code in `?e` command in order to be able to replace flags in string like the `?e` command did. Only then I could have removed `?e` command group and kept from failing lots of regression tests.~~

Actually that does work! Will change all places where it'll work!

ret2libc replied:
Sorry, what do you mean? which places are you going to replace?

brightprogrammer replied:
I mean I'll remove `"` where it's present with the `echo` command and not needed (as your TODO states)

ret2libc commented on librz/core/cmd_descs/cmd_shell.yaml:
```diff
@@ -109,9 +109,13 @@ commands:
+      - name: strs
         type: RZ_CMD_ARG_TYPE_STRING
         flags: RZ_CMD_ARG_FLAG_ARRAY
+        optional: true
```
I think you can just use one single argument named `strs` with FLAG_ARRAY and optional:true.

ret2libc commented on librz/core/cmd_descs/cmd_shell.yaml:
```diff
@@ -187,3 +191,45 @@ commands:
+        summary: Show numeric version
+        cname: show_version_numeric
+        args: []
+      - name: verjson
+        summary: Show version info in JSON format
+        cname: show_version_json
+        args: []
```
remove this and add RZ_OUTPUT_MODE_JSON in `ver` command

ret2libc commented on librz/core/cmd/cmd_print.c:
```diff
@@ -732,10 +733,12 @@ RZ_API void rz_core_set_asm_configs(RzCore *core, char *arch, ut32 bits, int seg
-	// TODO: integrate with =h& and bg analysis/string/searches/..
-	rz_core_cmd_help(core, help_msg_amper);
-}
+/* unused so compiler shows warning */
```
remove if unused.

XVilka replied:
@brightprogrammer you forgot to remove these

ret2libc commented on librz/core/cmd/cmd_print.c:
```diff
@@ -14,21 +14,22 @@
-	"&&", "", "wait until all tasks are finished",
-	NULL
-};
+/* used by helpCmdTasks function */
```
remove if unused.

ret2libc commented on test/db/json/json1:
```diff
@@ -7,7 +7,7 @@
 /wj
 ?j 0x10
-?Vj
+%vj
```
```suggestion
verj
```

ret2libc commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,806 @@
+	}
+}
+
+/* keeping these functions because they can be useful in future */
```
remove if unused

ret2libc commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,806 @@
+	return RZ_CMD_STATUS_OK;
+}
+
+RZ_IPI RzCmdStatus rz_help_handler(RzCore *core, int argc, const char **argv) {
```
move this in `cmd.c`

ret2libc commented on subprojects/rizin-shell-parser/grammar.js:
you changed all the style in this file and it becomes hard for me to review and know what actually changed :( please re-push with just the actual changes you made

brightprogrammer replied:
I ran prettier on this :sweat_smile: Which formatter do you use?
Changes made : 
- `math_arged_stmt` used `?` as identifier for starting of command. Now we have a choice between `%` (`percent_mark_identifier`) and `math_mark_identifier` (a regex) and all of these accept optional arguments (don't know whether it should be like that but if it's working then what can we say :smile:)
```js
    percent_mark_identifier: ($) => "%",
    percent_percent_mark_identifier: ($) => "%%",
    math_mark_identifier: ($) => /%[a-zA-Z0-9]*[!@#$&\-.+=]*/,
    _math_arged_stmt: ($) =>
      prec.left(
        1,
        choice(
          seq(
            field("command", $.percent_mark_identifier),
            field("args", optional($.args))
          ),
          seq(
            field("command", $.percent_percent_mark_identifier),
            field("args", optional($.args))
          ),
          seq(
            field("command", alias($.math_mark_identifier, $.cmd_identifier)),
            field("args", optional($.args))
          )
        )
      ),
```
- `_env_stmt_identifier` used a choice between `%` or `env`. Now it's just the `env` identifier.
```js
_env_stmt_identifier: ($) => "env",
```

Sorry for all the trouble :sweat_smile:

brightprogrammer replied:
I also changed the `simple_commands.txt` and `special_commands.txt`. You should check in `simple_commands.txt` for changed tests.

ret2libc reviewed (commented):
Terrific work! That's a big change and a very useful one, thanks!

Just few things missing to adjust, but overall things look good, good job!

XVilka commented:
```
../librz/core/cmd/cmd_shell.c:379: error: identifier expected
```
```
librz/core/cmd/cmd_shell.c:357:45: error: code should be clang-formatted [-Wclang-format-violations]
        rz_cons_printf("plug.init = %" PFMT64d "\n"
                                                   ^
librz/core/cmd/cmd_shell.c:358:36: error: code should be clang-formatted [-Wclang-format-violations]
                                   "plug.load = %" PFMT64d "\n"
                                                               ^
librz/core/cmd/cmd_shell.c:377:8: error: code should be clang-formatted [-Wclang-format-violations]
        switch(state->mode) {
              ^
librz/core/cmd/cmd_shell.c:387:5: error: code should be clang-formatted [-Wclang-format-violations]
                PJ* pj = state->d.pj;
                  ^
librz/core/cmd/cmd_shell.c:387:6: error: code should be clang-formatted [-Wclang-format-violations]
                PJ* pj = state->d.pj;
                   ^
```

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,597 @@
+	}
+}
+
+/// If third parameter (PJ) is NULL then standard output will be displayed
```
Convert into Doxygen

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,597 @@
+/// If third parameter (PJ) is NULL then standard output will be displayed
+/// If it's non-NULL then it's treated as a borrowd PJ and used and then
+/// printing will be done in JSON format and not standard one.
+/// So basically it's the third paramter that decides in which paramter
```
parameter

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,597 @@
+}
+
+/// If third parameter (PJ) is NULL then standard output will be displayed
+/// If it's non-NULL then it's treated as a borrowd PJ and used and then
```
borrowed

XVilka commented on librz/core/cmd_descs/cmd_descs.h:
```diff
@@ -4,11 +4,9 @@
 // WARNING: This file was auto-generated by cmd_descs_generate.py script. Do not
 // modify it manually. Look at cmd_descs.yaml if you want to update commands.
 //
-
```
This change is completely unnecessary, caused by the removal of some empty strings in the python script below

XVilka reviewed (changes requested):
A few more things left, before it's ready to be merged

XVilka commented on librz/core/cmd_descs/cmd_descs_generate.py:
```diff
@@ -47,14 +44,11 @@
 // WARNING: This file was auto-generated by cmd_descs_generate.py script. Do not
 // modify it manually. Look at cmd_descs.yaml if you want to update commands.
 //
-
```
You forgot this part - otherwise generated header/C file will be out-of-sync

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,603 @@
+}
+
+/**
+ * \brief If @p pj is NULL then standard output will be displayed
```
`@p` -> `\p`

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,603 @@
+ * printing will be done in JSON format and not standard one.
+ * Hence the the third parameter decides in which format data will be displayed.
+ *
+ * \param core RzCore*.
```
Just `RzCore`, no need to add pointer character

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,603 @@
+ *
+ * \param core RzCore*.
+ * \param input Input mathematical expression.
+ * \param pj Borrowed PJ* if calculated value is to be printed in JSON format.
```
Same

XVilka commented on librz/core/cmd/cmd_math.c:
```diff
@@ -0,0 +1,603 @@
+ * \param input Input mathematical expression.
+ * \param pj Borrowed PJ* if calculated value is to be printed in JSON format.
+ * \return true on success, false otherwise.
+ * */
```
Unnecessary space between `* *`

ret2libc commented on librz/core/cmd_descs/cmd_descs_generate.py:
```diff
@@ -29,9 +29,7 @@
 //
 
 #include <cmd_descs.h>
-
```
keep these, they don't hurt and they help splitting the output.

ret2libc commented on librz/core/cmd_descs/cmd_descs_generate.py:
```diff
@@ -29,9 +29,7 @@
 #include <cmd_descs.h>
-
 {helps_declarations}
-
```
also this

ret2libc commented on librz/core/cmd_descs/cmd_descs_generate.py:
```diff
@@ -54,7 +52,6 @@
 
 // Command handlers, manually defined somewhere else
 {handlers_declarations}
-
```
and this

ret2libc reviewed (approved):
Few wrong help messages (e.g. missing spaces) in `%?` but that's fine, let's do separate PR. Good job! :)
